### DISCUSSION TITLE
MAC restriction in Strongswan possible?

### DISCUSSION THREAD

**Discussion by yoloknight** [strongswan/strongswan#402] [Q&A] [answered]
Hi,

is it possible to restrict the traffic inside the vpn/ipsec via MAC? (like in iptables)
Like: I only accpet the traffic from MAC xy and all other packets are droped.

Or deactivate other interface i.e. vpn via caple and swanctl blocks the wlan interface? (i think this more likely a script thing and not in particular strongswan, but it would be interessing if its possible)

If you have expierence with simmilar soloutions i would be glad to hear it. :)

regards

**Comment by tobiasbrunner** [comment, marked as answer]
I guess you could mark packets via `ebtables` based on the MACs and then use the marks in `iptables` and/or the IPsec SAs/policies. Something similar could also be done for packets from specific interfaces (but you could also use an updown script to block traffic from certain interfaces completely).

**Reply by yoloknight** [reply]
Thanks that would worke for me. :)

Well I have also another question about swanctl for the autostart of a IPSec connection. Is it possible that the swanctl-charon service detect itself a new network connection and try to establish the IKE with the gateway?

I tried it, but it dosent work (also i have close_action and start_action both on 'start'). I noticed if I suspend my Laptop it wont bring the connection up. Is there a way to configure a behavior that the swanctl service try constantly the restart the IKE regardless which interface, so long the IKE is be able to establish? (i use a roadwarrior connection)

**Reply by tobiasbrunner** [reply]
There are several settings that might help (e.g. DPD and MOBIKE or `keyingtries` or even using trap policies so SAs are automatically (re-)established on matching outbound traffic). Device suspension is particularly tricky as the daemon (and kernel, e.g. for lifetimes of SAs) usually uses a monotonic clock, which does not increase while the device is suspended. So after waking up, the daemon does not really notice that there was an interruption. Depending on the kernel events and whether the local IP changed there might be a MOBIKE update, which updates the SAs or would reestablish them if the server already cleared its state, but that's not necessarily the case. However, the `charon.check_current_path` option can be used in newer releases to force a DPD even if the local IP has not changed (still requires events from the kernel e.g. for addresses getting added or interfaces coming up to trigger it). You'd have to check in the log what exactly happens in such cases.

**Reply by yoloknight** [reply]
Thanks again for the quick reply :)
DPD and MOBIKE had I activated but the charon.check_current_path Im not sure if it works.

I edited in /etc/strongswan.d the charon.conf and added the "check_current_path = yes", but no changes at all. Did I do something wrong about it?

For the trap policies are there some examples on the Strongswan page?

**Reply by tobiasbrunner** [reply]
You need at least 5.9.0. And it only has an effect when a roam event is triggered but the IP stays the same (read the log to see if any of that happens). To use traps, you basically just have to configure `start_action=trap`.

**Reply by yoloknight** [reply]
Hi Tobias,

I tried some of the elements what you suggest for me. Setup is a Laptop to a Wlan-Router and the Strongswan swanctl 5.8.2 (cant go above at the moment, when will 5.9 in generell via apt-get install published?). 


I had the following results:

MOIKE, DPD, Startaction = start, Stopaction = start, keyingries = 0 ---> Tunnel starts if the WLAN Connection to the Router is accessable. But if you have no Wlan Connection from the start and only later connect to the WLAN Router, the Tunnel is not established. (its normal for the Modi 'start' I guess)
But if I change Start_action = trap, Stop_action = trap, there is no automatic connection to a WLAN Router. (did I understand trap wrong?)

Are there example scripts to implement a behaivor to force the IKE Tunnel no matter what? (i.e. can i write a script for swanctl to force dpd with simple commands or do I have do more for it?

My use case is:
Boot -> if wlan/lan there...build the Tunnel
Boot + no wlan -> login wlan -> if wlan/lan there...build the tunnel
connection loss or suspend -> swanctl should check if it can build the tunnel

I would try as next step your suggestion with check_current_path in 5.9, but it would be nice if I can reliable way to force swantcl to build the tunnel no matter what. 

Sorry for that many questions. :)

**Reply by yoloknight** [reply]
This is the message I get for the Trap. What I am doing wrong?

Mai 20 14:51:38 host charon[969]: 15[IKE] giving up after 5 retransmits
Mai 20 14:51:38 host charon[969]: 15[IKE] peer not responding, trying again (4/0)
Mai 20 14:51:38 host charon[969]: 15[IKE] initiating IKE_SA ipsec[1] to 20.20.20.10
Mai 20 14:51:38 host charon[969]: 15[IKE] initiating IKE_SA ipsec[1] to 20.20.20.10
Mai 20 14:51:38 host charon[969]: 15[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
Mai 20 14:51:38 host charon[969]: 15[NET] sending packet: from 192.169.1.2[500] to 20.20.20.10[500] (464 bytes)
Mai 20 14:51:39 host charon[969]: 05[KNL] creating delete job for CHILD_SA ESP/0x00000000/20.20.20.10
Mai 20 14:51:39 host charon-systemd[943]: creating delete job for CHILD_SA ESP/0x00000000/20.20.20.10
Mai 20 14:51:39 host charon[969]: 05[JOB] CHILD_SA ESP/0x00000000/20.20.20.10 not found for delete
Mai 20 14:51:39 host charon-systemd[943]: CHILD_SA ESP/0x00000000/20.20.20.10 not found for delete
Mai 20 14:51:39 host charon[969]: 07[KNL] creating delete job for CHILD_SA ESP/0x00000000/20.20.20.10
Mai 20 14:51:39 host charon-systemd[943]: creating delete job for CHILD_SA ESP/0x00000000/20.20.20.10
Mai 20 14:51:39 host charon[969]: 07[JOB] CHILD_SA ESP/0x00000000/20.20.20.10 not found for delete
Mai 20 14:51:39 host charon-systemd[943]: CHILD_SA ESP/0x00000000/20.20.20.10 not found for delete
Mai 20 14:51:41 host charon[969]: 09[KNL] creating acquire job for policy 192.169.1.2/32[udp/50428] === 208.123.73.4/32[udp] with reqid {2}
Mai 20 14:51:41 host charon-systemd[943]: creating acquire job for policy 192.169.1.2/32[udp/50428] === 208.123.73.4/32[udp] with reqid {2}
Mai 20 14:51:41 host charon[969]: 09[CFG] ignoring acquire, connection attempt pending
Mai 20 14:51:41 host charon-systemd[943]: trap not found, unable to acquire reqid 2

**Reply by tobiasbrunner** [reply]
Note that according to the log, you have two IKE daemons running (charon and charon-systemd), you definitely want to avoid that.

> But if you have no Wlan Connection from the start and only later connect to the WLAN Router, the Tunnel is not established.

Read the log for details. Check if there are retries or what else is going on. There is also an option (_charon.retry_initiate_interval_) that enables retries if DNS resolution doesn't work initially (only applies if you use a FQDN for the server address).

> But if I change Start_action = trap, Stop_action = trap, there is no automatic connection to a WLAN Router. (did I understand trap wrong?)

It's triggered by traffic to the remote traffic selector (if you use `0.0.0.0/0`, any packet should trigger an acquire). Also, you should configure _keyingtries=1_ (the default) and set _dpd_action_ and _close_action_ to _clear_ (also the default) if you use this as another acquire will be triggered after the retransmission timeout (otherwise, it could lead to duplicates).

> Are there example scripts to implement a behaivor to force the IKE Tunnel no matter what?

@Thermi wrote one (see [this repository](https://gitlab.com/Thermi/ipsecTunnelMonitoringScript)). But I don't know exactly if it is applicable in this scenario.

> I would try as next step your suggestion with check_current_path in 5.9, but it would be nice if I can reliable way to force swantcl to build the tunnel no matter what.

That's an option for the suspension case, where established SAs continue to exist, but they might not be usable because the server or NAT device already removed state. The option triggers a DPD to detect this situation and either update the existing SAs or recreate them. You could theoretically replicate this (with more overhead) from a script by triggering a rekeying (there is no command to trigger a DPD).

**Reply by yoloknight** [reply]
Hi Tobias,
thanks again for your quick reply. So far as I tested the problems with the DPD detection are with our Fortigate.
This Thead can be blosed. :)
